### PR TITLE
HydratedRuntime: Chapter One

### PR DESCRIPTION
**Pull request by pinkiebell**
- Extensible contract for computing proofs.
- Refactor -/ DRY'ing of code
- Proper support for compact memory proofs -/ sparse memory
- Reduced contract size and gas usage for EVMRuntime

### CODE REVIEW

**Comment by pinkiebell** [comment]
Hold on, I will solve it differently. The current solutions bugs me a bit :)

File: .solcover.js [hunk 1]
```diff
@@ -3,5 +3,5 @@ module.exports = {
   testCommand: '../node_modules/.bin/truffle test --network coverage test/contracts/*',
   norpc: true,
   deepSkip: true,
-  skipFiles: ['EthereumRuntime.sol', 'IEthereumRuntime.sol', 'CompactEthereumRuntime.sol', 'EVMRuntime.sol'],
+  skipFiles: ['CompactEthereumRuntime.sol', 'EVMRuntime.sol'],
```
Comment (johannbarbie) [inline .solcover.js]: `CompactEthereumRuntime.sol` is completely duplicated effort, right? why don't we get rid of it completely?

File: contracts/HydratedRuntime.sol [hunk 2]
```diff
@@ -0,0 +1,103 @@
+pragma solidity 0.5.2;
+pragma experimental ABIEncoderV2;
+
+
+import "./EVMCode.slb";
+import "./EVMStack.slb";
+import "./EVMMemory.slb";
+import "./EVMAccounts.slb";
+import "./EVMLogs.slb";
```
Comment (johannbarbie) [inline contracts/HydratedRuntime.sol]: no need to import these if already imported by inherited class (`EVMRuntime`)

**Comment by pinkiebell** [comment]
> > test/contracts/TestEVMMemory.sol
> 
> shouldn't we have a replacement for these? or do the fixtures cover everything?

Functionality is covered. But we loose our asserts, I will add fixtures who will throw and support that in the tests

**Comment by pinkiebell** [comment]
> > > test/contracts/TestEVMMemory.sol
> > 
> > 
> > shouldn't we have a replacement for these? or do the fixtures cover everything?
> 
> Functionality is covered. But we loose our asserts, I will add fixtures who will throw and support that in the tests

Just checked. We actually didn't test the asserts there in the first place.
But I will still do it in a upcoming pr

**Pull request merged**